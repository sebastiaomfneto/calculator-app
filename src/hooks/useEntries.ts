import { Reducer, useReducer } from 'react';

import { calculateExpression } from '../helpers';

type State = {
  entries: string;
  result: string;
};

type Action =
  | { type: 'ADD_ENTRY'; payload: string }
  | { type: 'CALCULATE_RESULT' }
  | { type: 'RESET' };

export const ADD_ENTRY = 'ADD_ENTRY';
export const CALCULATE_RESULT = 'CALCULATE_RESULT';
export const RESET = 'RESET';

const reducer: Reducer<State, Action> = (
  prevState: State,
  action: Action
): State => {
  switch (action.type) {
    case ADD_ENTRY:
      return {
        entries: prevState.entries.concat(action.payload),
        result: '',
      };
    case CALCULATE_RESULT:
      return {
        entries: '',
        result: calculateExpression(prevState.entries),
      };
    case RESET:
      return {
        entries: '',
        result: '',
      };
    default:
      throw new Error('Not implemented');
  }
};

export const useEntries = (
  initialState: State = { entries: '', result: '' }
) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return {
    state,
    dispatch,
  };
};
