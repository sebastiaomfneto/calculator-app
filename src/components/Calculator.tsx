import { FC, useMemo } from 'react';
import {
  Paper,
  Grid,
  GridProps,
  Button,
  ButtonProps,
  Box,
  Typography,
  IconButton,
} from '@material-ui/core';
import { Backspace } from '@material-ui/icons';

import { useEntries, ADD_ENTRY, CALCULATE_RESULT, RESET } from '../hooks';

type CalculatorButton = Pick<
  ButtonProps & GridProps,
  'color' | 'variant' | 'onClick' | 'xs'
> & {
  value: string;
};

export const Calculator: FC = () => {
  const {
    state: { entries, result },
    dispatch,
  } = useEntries();

  const buttons = useMemo<CalculatorButton[]>(
    () => [
      {
        value: '7',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '7' }),
      },
      {
        value: '8',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '8' }),
      },
      {
        value: '9',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '9' }),
      },
      {
        value: '/',
        color: 'primary',
        variant: 'outlined',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '/' }),
      },
      {
        value: '4',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '4' }),
      },
      {
        value: '5',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '5' }),
      },
      {
        value: '6',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '6' }),
      },
      {
        value: '*',
        color: 'primary',
        variant: 'outlined',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '*' }),
      },
      {
        value: '1',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '1' }),
      },
      {
        value: '2',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '2' }),
      },
      {
        value: '3',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '3' }),
      },
      {
        value: '-',
        color: 'primary',
        variant: 'outlined',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '-' }),
      },
      {
        value: '0',
        color: 'default',
        variant: 'text',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '0' }),
      },
      {
        value: '=',
        color: 'primary',
        variant: 'contained',
        xs: 6,
        onClick: () => dispatch({ type: CALCULATE_RESULT }),
      },

      {
        value: '+',
        color: 'primary',
        variant: 'outlined',
        xs: 3,
        onClick: () => dispatch({ type: ADD_ENTRY, payload: '+' }),
      },
    ],
    [dispatch]
  );

  return (
    <Paper variant="outlined">
      <Box padding={3}>
        <Box marginY={1}>
          <Paper variant="outlined">
            <Box
              display="flex"
              alignItems="center"
              justifyContent="flex-end"
              gridColumnGap="8px"
              height="56px"
              paddingY={2}
              paddingX={1}
            >
              <Typography align="right" data-testid="result-text">
                {result || entries || '0'}
              </Typography>
              <IconButton
                size="small"
                onClick={() => dispatch({ type: RESET })}
                data-testid="button-reset"
              >
                <Backspace fontSize="small" />
              </IconButton>
            </Box>
          </Paper>
        </Box>
        <Grid container spacing={1}>
          {buttons.map(({ color, variant, value, xs, onClick }) => (
            <Grid key={value} item xs={xs}>
              <Button
                color={color}
                variant={variant}
                fullWidth
                onClick={onClick}
                data-testid={`button-${value}`}
              >
                {value}
              </Button>
            </Grid>
          ))}
        </Grid>
      </Box>
    </Paper>
  );
};
