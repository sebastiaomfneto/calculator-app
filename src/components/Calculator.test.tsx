import { render, act, fireEvent } from '@testing-library/react';

import { Calculator } from './Calculator';
import * as hooks from '../hooks';

jest.mock('../hooks');

const renderWithHook = (Element: any, state: any, dispatch: any) => {
  jest
    .spyOn(hooks, 'useEntries')
    .mockImplementationOnce(() => ({ state, dispatch }));

  return render(Element);
};

describe('Calculator', () => {
  const state = { entries: '', result: '' };
  const dispatch = jest.fn();

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should match with snapshot', () => {
    const { container } = renderWithHook(<Calculator />, state, dispatch);

    expect(container).toMatchSnapshot();
  });

  test('should show 0 when state result & entries is empty', async () => {
    const { findByTestId } = renderWithHook(<Calculator />, state, dispatch);

    expect(await findByTestId('result-text')).toHaveTextContent('0');
  });

  test('should show state result', async () => {
    const { findByTestId } = renderWithHook(
      <Calculator />,
      { entries: '2+2', result: '4' },
      dispatch
    );

    expect(await findByTestId('result-text')).toHaveTextContent('4');
  });

  test('should show state entries', async () => {
    const { findByTestId } = renderWithHook(
      <Calculator />,
      { entries: '2+2', result: '' },
      dispatch
    );

    expect(await findByTestId('result-text')).toHaveTextContent('2+2');
  });

  test('should call dispatch "ADD_ENTRY" when number & operations button clicked', async () => {
    const { findByTestId } = renderWithHook(<Calculator />, state, dispatch);

    await act(async () => {
      fireEvent.click(await findByTestId('button-2'));
      fireEvent.click(await findByTestId('button-+'));
      fireEvent.click(await findByTestId('button-2'));
    });

    expect(dispatch).toHaveBeenCalledWith({
      type: hooks.ADD_ENTRY,
      payload: '2',
    });
    expect(dispatch).toHaveBeenCalledWith({
      type: hooks.ADD_ENTRY,
      payload: '+',
    });
    expect(dispatch).toHaveBeenCalledWith({
      type: hooks.ADD_ENTRY,
      payload: '2',
    });
  });
  test('should call dispatch "CALCULATE_RESULT" when equals button clicked', async () => {
    const { findByTestId } = renderWithHook(<Calculator />, state, dispatch);

    await act(async () => {
      fireEvent.click(await findByTestId('button-='));
    });

    expect(dispatch).toHaveBeenCalledWith({
      type: hooks.CALCULATE_RESULT,
    });
  });
  test('should call dispatch "RESET" when clear button clicked', async () => {
    const { findByTestId } = renderWithHook(<Calculator />, state, dispatch);

    await act(async () => {
      fireEvent.click(await findByTestId('button-reset'));
    });

    expect(dispatch).toHaveBeenCalledWith({
      type: hooks.RESET,
    });
  });
});
