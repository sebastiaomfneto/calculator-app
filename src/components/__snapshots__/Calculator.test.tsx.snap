// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should match with snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-3"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignRight"
              data-testid="result-text"
            >
              0
            </p>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              data-testid="button-reset"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 3H7c-.69 0-1.23.35-1.59.88L0 12l5.41 8.11c.36.53.9.89 1.59.89h15c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-3 12.59L17.59 17 14 13.41 10.41 17 9 15.59 12.59 12 9 8.41 10.41 7 14 10.59 17.59 7 19 8.41 15.41 12 19 15.59z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-7"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              7
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-8"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              8
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-9"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              9
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
            data-testid="button-/"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              /
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-4"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              4
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-5"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              5
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-6"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              6
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
            data-testid="button-*"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              *
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-1"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              1
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-2"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              2
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-3"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              3
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
            data-testid="button--"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              -
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
            data-testid="button-0"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              0
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
            data-testid="button-="
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              =
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
            data-testid="button-+"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              +
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
